File: src/types/asset.ts
export type AssetType = 'sample' | 'midi' | 'preset' | 'project';
export type IndexStatus = 'pending' | 'indexed' | 'missing' | 'error';
export type KeyScale = 'major' | 'minor';
export type BpmSource = 'filename' | 'audio_analysis' | 'midi';

export interface MidiNote {
  pitch: number;
  startTick: number;
  durationTicks: number;
  velocity: number;
}

export interface SampleMeta {
  bitDepth?: number;
}

export interface MidiMeta {
  timeSignature?: string;
  barCount?: number;
  noteCount?: number;
  pianoRoll?: MidiNote[];
  tracks?: number;
  noteRangeLow?: number;
  noteRangeHigh?: number;
}

export interface PresetMeta {
  synth: string;
  category: string;
  tags: string[];
}

/** A single clip placed on the FL Studio playlist. */
export interface PlaylistClip {
  /** Playlist track index (top-down, 0-based). */
  track: number;
  /** Position on the timeline, in PPQ ticks. */
  positionTicks: number;
  /** Visible clip length on the timeline, in PPQ ticks. */
  lengthTicks: number;
  /** Trim from the source start, in PPQ ticks. */
  startOffsetTicks: number;
  /** Trim from the source end, in PPQ ticks. */
  endOffsetTicks: number;
  /** Set when the clip references a pattern (0-based index into `patterns`). */
  patternIndex: number | null;
  /** Set when the clip references an audio/automation channel. */
  channelIndex: number | null;
  muted: boolean;
}

export interface ProjectMeta {
  daw: string;
  version?: string;
  trackCount?: number;
  tempo?: number;
  timeSignature?: string;
  lastModified?: number;
  plugins?: string[];
  sampleCount?: number;
  /** Sample / audio file paths referenced by the project, when extractable. */
  samples?: string[];
  title?: string;
  author?: string;
  genre?: string;
  comments?: string;
  fileSizeBytes?: number;
  /** Channel (instrument) names from the project rack. */
  channels?: string[];
  /** Pattern names defined in the project. */
  patterns?: string[];
  /** Mixer-insert names (FL Studio "Mixer tracks", distinct from playlist tracks). */
  mixerTracks?: string[];
  url?: string;
  /** Pulses per quarter note from FLhd. Required to convert clip ticks → beats. */
  ppq?: number;
  /** Playlist clips parsed from the FLP arrangement. Mirrors `arrangements[0].clips`. */
  clips?: PlaylistClip[];
  /** One entry per FL Studio arrangement (FL 12.9+). Always populated. */
  arrangements?: Arrangement[];
}

/** A single FL Studio arrangement (named playlist). */
export interface Arrangement {
  index: number;
  name: string | null;
  clips: PlaylistClip[];
}

export type AssetMeta = SampleMeta | MidiMeta | PresetMeta | ProjectMeta | Record<string, unknown>;

export interface Asset {
  id: string;
  path: string;
  filename: string;
  extension: string;
  type: AssetType;
  packId: string | null;
  packName: string | null;

  bpm: number | null;
  keyNote: string | null;
  keyScale: KeyScale | null;
  durationMs: number | null;
  sampleRate: number | null;
  channels: number | null;

  instrument: string | null;
  subtype: string | null;

  /** Smart tags derived from filename analysis */
  energyLevel: 'high' | 'low' | null;
  texture: 'organic' | 'synthetic' | null;
  space: 'dry' | 'wet' | null;
  role: 'foundation' | 'top_end' | 'ear_candy' | null;

  isFavorite: boolean;
  userTags: string[];
  playCount: number;
  lastPlayed: number | null;
  rating: number | null;

  meta: AssetMeta;

  indexStatus: IndexStatus;
  waveformData: number[] | null;
  bpmSource: BpmSource | null;
  keySource: BpmSource | null;

  createdAt: number;
  updatedAt: number;
}

export interface SearchResult {
  assets: Asset[];
  total: number;
}

File: src-tauri/src/models/asset.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Asset {
    pub id: String,
    pub path: String,
    pub filename: String,
    pub extension: String,
    #[serde(rename = "type")]
    pub asset_type: String,
    pub pack_id: Option<String>,
    pub pack_name: Option<String>,

    pub bpm: Option<f32>,
    pub key_note: Option<String>,
    pub key_scale: Option<String>,
    pub duration_ms: Option<i64>,
    pub sample_rate: Option<i64>,
    pub channels: Option<i64>,

    pub instrument: Option<String>,
    pub subtype: Option<String>,

    pub is_favorite: bool,
    pub user_tags: Vec<String>,
    pub play_count: i64,
    pub last_played: Option<i64>,
    pub rating: Option<i64>,

    pub meta: serde_json::Value,

    pub index_status: String,
    pub waveform_data: Option<Vec<f32>>,
    pub bpm_source: Option<String>,
    pub key_source: Option<String>,

    /// Smart tags derived from filename analysis
    pub energy_level: Option<String>,
    pub texture: Option<String>,
    pub space: Option<String>,
    pub role: Option<String>,

    pub created_at: i64,
    pub updated_at: i64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SearchQuery {
    pub filters: AssetFilters,
    pub sort: SortOptions,
    #[serde(default = "default_limit")]
    pub limit: i64,
    #[serde(default)]
    pub offset: i64,
}

fn default_limit() -> i64 {
    100
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct AssetFilters {
    #[serde(default)]
    pub query: String,
    #[serde(default)]
    pub types: Vec<String>,
    #[serde(default)]
    pub pack_ids: Vec<String>,
    #[serde(default)]
    pub instruments: Vec<String>,
    #[serde(default)]
    pub subtypes: Vec<String>,
    pub bpm_min: Option<f32>,
    pub bpm_max: Option<f32>,
    #[serde(default)]
    pub keys: Vec<String>,
    #[serde(default)]
    pub scales: Vec<String>,
    #[serde(default)]
    pub favorites_only: bool,
    #[serde(default)]
    pub tags: Vec<String>,
    pub path_prefix: Option<String>,
    /// Smart tag filters
    #[serde(default)]
    pub energy_levels: Vec<String>,
    #[serde(default)]
    pub textures: Vec<String>,
    #[serde(default)]
    pub spaces: Vec<String>,
    #[serde(default)]
    pub roles: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SortOptions {
    pub field: String,
    pub direction: String,
}

impl Default for SortOptions {
    fn default() -> Self {
        Self {
            field: "filename".into(),
            direction: "asc".into(),
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SearchResult {
    pub assets: Vec<Asset>,
    pub total: i64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MidiNote {
    pub pitch: u8,
    pub start_tick: u32,
    pub duration_ticks: u32,
    pub velocity: u8,
}

/// A single facet bucket: value + count.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FacetCount {
    pub value: String,
    pub count: i64,
}

/// All facet counts returned in one IPC call.
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct FacetCounts {
    pub instruments: Vec<FacetCount>,
    pub subtypes: Vec<FacetCount>,
    pub energy_levels: Vec<FacetCount>,
    pub textures: Vec<FacetCount>,
    pub spaces: Vec<FacetCount>,
    pub roles: Vec<FacetCount>,
}
